src/runtime/hashmap.go

```diff
@@ -83,7 +116,7 @@ func hashmapLenUnsafePointer(p unsafe.Pointer) int {
 
 // Set a specified key to a given value. Grow the map if necessary.
 //go:nobounds
-func hashmapSet(m *hashmap, key unsafe.Pointer, value unsafe.Pointer, hash uint32, keyEqual func(x, y unsafe.Pointer, n uintptr) bool) {
+func hashmapSet(m *hashmap, key unsafe.Pointer, value unsafe.Pointer, hash uint32, keyEqual func(x, y unsafe.Pointer, n uintptr) bool, keyHash func(y unsafe.Pointer, size uintptr) uint32) {
```

Comment:
I had the thought of using `hashmapAlgorithms[m.alg]` entries as needed, instead of passing functions pointers around to the various functions that need them.

Reply:
Yes, that would probably be better.
The optimizer has a lot of trouble looking through function pointers, so it's generally a good idea to get rid of them. Using `hashmapAlgorithms[alg]` still is a function pointer though: it's unlikely the optimizer manages to look through it. The most efficient way (although perhaps not the most readable way) is to use a switch:

```go
func hashmapEqual(x, y unsafe.Pointer, n uintptr, alg int) bool {
	switch alg {
	case hashmapAlgorithmString:
		return hashmapStringEqual(x, y, n)
	case hashmapAlgorithmBinary:
		// ..etc
}
```

Regular function calls still allow inlining and various other interprocedural optimizations such as const propagations.

Reply:
Ok, I'll fix up the code to rely on the passed `alg` value and switches.